google_guest_agent/telemetry/telemetry.go

```diff
@@ -0,0 +1,100 @@
+//  Copyright 2023 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package telemetry
+
+import (
+	"context"
+	"encoding/base64"
+	"net/http"
+	"time"
+
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+	"google.golang.org/protobuf/proto"
+
+	tpb "github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/telemetry/proto"
+)
+
+var (
+	metadataURL = "http://169.254.169.254/computeMetadata/v1/"
+)
+
+// Data is telemetry data on the current agent and OS.
+type Data struct {
+	// Name of the agent.
+	AgentName string
+	// Version of the Agent.
+	AgentVersion string
+	// Architecture of the Agent.
+	AgentArch string
+
+	// OS name.
+	OS string
+	// The name the OS uses to fully describe itself.
+	LongName string
+	// OS name in short form (aka distro name).
+	ShortName string
+	// Version of the OS.
+	Version string
+	// Kernel Release.
+	KernelRelease string
+	// Kernel Version.
+	KernelVersion string
+}
+
+func formatGuestAgent(d Data) string {
+	data, err := proto.Marshal(&tpb.AgentInfo{
+		Name:         &d.AgentName,
+		Version:      &d.AgentVersion,
+		Architecture: &d.AgentArch,
+	})
+	if err != nil {
+		logger.Warningf("Error marshalling AgentInfo: %v", err)
+	}
+	return base64.StdEncoding.EncodeToString(data)
+}
+
+func formatGuestOS(d Data) string {
+	data, err := proto.Marshal(&tpb.OSInfo{
+		OsType:        &d.OS,
+		LongName:      &d.LongName,
+		ShortName:     &d.ShortName,
+		Version:       &d.Version,
+		KernelVersion: &d.KernelVersion,
+		KernelRelease: &d.KernelRelease,
+	})
+	if err != nil {
+		logger.Warningf("Error marshalling AgentInfo: %v", err)
+	}
+	return base64.StdEncoding.EncodeToString(data)
+}
+
+// Record records telemetry data.
+func Record(ctx context.Context, d Data) error {
+	client := &http.Client{
+		Timeout: 10 * time.Second,
+	}
+
+	req, err := http.NewRequest("GET", metadataURL, nil)
```

Comment:
Can you move this mds call to the metadata package and just wrap it from here? You can reuse the metadata client in the `main.go` passing it down to this function - later on I plan to improve the call pattern of the mds client but for now it should be enough.

Reply:
What part should I wrap? I need to add headers, should I update the existing Get call to add optional headers? I don't actually care about metadata in this function at all (or any of the other code in that function), just need to make some call that includes these headers, updating Get seemed more complicated as that impacted any other callers, also it would include an unneeded json unmarhal.